### PR TITLE
MB-8054 Exploring using open telemetry for metrics and tracing

### PR DESCRIPTION
**Pull request by ahobson**
## Description

This takes some the work from Nelz and Chas and tries to apply it to MilMove.

It also switches to using gorilla mux so we can get some better tagging of request stats.

## Reviewer Notes

The telemetry is disabled by default and the metrics we are gathering still needs some work, but this seems like a decent foundation to build on.

## Setup

Add any steps or code to run in this section to help others prepare to run your code:

```sh
echo "Code goes here"
```

## Code Review Verification Steps

* [ ] If the change is risky, it has been tested in experimental before merging.
* [ ] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/wiki/Backend-Programming-Guide#logging)
* [ ] The requirements listed in [Querying the Database Safely](https://github.com/transcom/mymove/wiki/Backend-Programming-Guide#querying-the-database-safely) have been satisfied.
* Any new migrations/schema changes:
  * [ ] Follow our guidelines for zero-downtime deploys (see [Zero-Downtime Deploys](https://github.com/transcom/mymove/wiki/migrate-the-database#zero-downtime-migrations))
  * [ ] Have been communicated to #g-database
  * [ ] Secure migrations have been tested following the instructions in our [docs](https://github.com/transcom/mymove/wiki/migrate-the-database#secure-migrations)
* [ ] There are no aXe warnings for UI.
* [ ] This works in [Supported Browsers and their phone views](https://github.com/transcom/mymove/tree/master/docs/adr/0016-Browser-Support.md) (Chrome, Firefox, IE, Edge).
* [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
* [ ] User facing changes have been reviewed by design.
* [ ] Request review from a member of a different team.
* [ ] Have the Jira acceptance criteria been met for this change?

## References

* [Jira story](tbd) for this change
* [this article](tbd) explains more about the approach used.

## Screenshots

If this PR makes visible UI changes, an image of the finished UI can help reviewers and casual
observers understand the context of the changes. A before image is optional and
can be included at the submitter's discretion.

Consider using an animated image to show an entire workflow instead of using multiple images. You may want to use GIPHY CAPTURE for this! 📸

_Please frame screenshots to show enough useful context but also highlight the affected regions._

### CODE REVIEW

**Review by felipe-lee — COMMENTED** [review]
I've looked over the files and agree that I think it should be fine to merge and deploy without it actually turning on anywhere except experimental. I left one comment about the docker setup, but haven't actually tested out the docker setup, only tested it with stdout so far. One thing I think would be good before merging would be at least a small blurb doc-wise about how to run this (like basically the rundown you gave us in the meeting about env vars and the docker setup). I know it'll possibly change as we mess with it more, but I think if it's going to be in master, it should have some accompanying docs. I'll mess with running it in docker tomorrow, but at least what I'm seeing in stdout is really cool (if a bit messy to read in stdout haha).

I haven't had a chance to dig into more best practices for using otel, but since the point of this pr is more to get it working, I think we can bring those in as we mess with it more.

**Comment by ahobson** [comment]
> I've looked over the files and agree that I think it should be fine to merge and deploy without it messing anything up or actually turning on anywhere except experimental. I left one comment about the docker setup, but haven't actually tested out the docker setup, only tested it with stdout so far.
> 
> One thing I think would be good before merging would be at least a small blurb doc-wise about how to run this (like basically the rundown you gave us in the meeting about env vars and the docker setup). I know it'll possibly change as we mess with it more, but I think if it's going to be in master, it should have some accompanying docs.

Where do you think those docs should be? I was thinking this would be something to add to Docusaurus?

File: config/telemetry/collector-config.dev.yaml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    protocols:
+      grpc:
+        endpoint: 0.0.0.0:55680
+  hostmetrics:
+    collection_interval: 1m
+    scrapers:
+      load:
+      memory:
+
+processors:
```
[the remark this answers is not in the document]
Comment (ahobson) [reply config/telemetry/collector-config.dev.yaml:12]: added!

**Comment by ahobson** [comment]
> > I've looked over the files and agree that I think it should be fine to merge and deploy without it messing anything up or actually turning on anywhere except experimental. I left one comment about the docker setup, but haven't actually tested out the docker setup, only tested it with stdout so far.
> > One thing I think would be good before merging would be at least a small blurb doc-wise about how to run this (like basically the rundown you gave us in the meeting about env vars and the docker setup). I know it'll possibly change as we mess with it more, but I think if it's going to be in master, it should have some accompanying docs.
> 
> Where do you think those docs should be? I was thinking this would be something to add to Docusaurus?

For a quick test, here are the instructions:
1. In one window run `docker-compose -f docker-compose.telemetry.yml up`
2. In another, run `TELEMETRY_ENDPOINT=localhost:13133 TELEMETRY_ENABLED=1 make server_run`
3. Visit http://locahost:5601

**Comment by felipe-lee** [comment]
> > > I've looked over the files and agree that I think it should be fine to merge and deploy without it messing anything up or actually turning on anywhere except experimental. I left one comment about the docker setup, but haven't actually tested out the docker setup, only tested it with stdout so far.
> > > One thing I think would be good before merging would be at least a small blurb doc-wise about how to run this (like basically the rundown you gave us in the meeting about env vars and the docker setup). I know it'll possibly change as we mess with it more, but I think if it's going to be in master, it should have some accompanying docs.
> > 
> > 
> > Where do you think those docs should be? I was thinking this would be something to add to Docusaurus?
> 
> For a quick test, here are the instructions:
> 
>     1. In one window run `docker-compose -f docker-compose.telemetry.yml up`
> 
>     2. In another, run `TELEMETRY_ENDPOINT=localhost:13133 TELEMETRY_ENABLED=1 make server_run`
> 
>     3. Visit http://locahost:5601


Yeah I think Docusaurus would be a good spot for them.

That sounds like a good snippet and we can expand on it as we build it out more.

**Comment by ahobson** [comment]
> > ```
> > 1. In one window run `docker-compose -f docker-compose.telemetry.yml up`
> > 
> > 2. In another, run `TELEMETRY_ENDPOINT=localhost:13133 TELEMETRY_ENABLED=1 make server_run`
> > 
> > 3. Visit http://locahost:5601
> > ```

https://transcom.github.io/mymove-docs/docs/dev/tools/telemetry/Running-telemetry-locally

Anything else you think we need to approve this PR?

**Review by felipe-lee — APPROVED** [review]
LGTM. I think this is at a good merging point and we can iron out exactly what we need as we work with it more.

**Pull request merged**